include/secp256k1_frost.h

```diff
@@ -0,0 +1,713 @@
+ */
+SECP256K1_API int secp256k1_frost_share_serialize(
+    const secp256k1_context* ctx,
+    unsigned char *out32,
+    const secp256k1_frost_share* share
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Parse a FROST share.
+ *
+ *  Returns: 1 when the share could be parsed, 0 otherwise.
+ *  Args:    ctx: a secp256k1 context object
+ *  Out:   share: pointer to a share object
+ *  In:     in32: pointer to the 32-byte share to be parsed
+ */
+SECP256K1_API int secp256k1_frost_share_parse(
+    const secp256k1_context* ctx,
+    secp256k1_frost_share* share,
+    const unsigned char *in32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Creates a key generation share and verifiable secret sharing ("VSS")
+ *  commitments.
+ *
+ *  To generate a key, each participant generates a share for each other
+ *  participant. For example, in the case of 2 particpants, Alice and Bob, they
+ *  each generate 2 shares, distribute 1 share to each other using a secure
+ *  channel, and keep 1 for themselves.
+ *
+ *  Each participant _must_ have a secure channel with each other participant
+ *  with which they can transmit shares to each other.
+ *
+ *  A new session_id32 _must_ be used for each key generation session. For
+ *  example, in the case of 2 participants, Alice and Bob, Alice will generate
+ *  a session_id32 and use it for each of the 2 calls to
+ *  secp256k1_frost_share_gen and Bob will generate a session_id32 and use it
+ *  for each of the 2 calls to secp256k1_frost_share_gen. Both Alice and Bob
+ *  must NOT REUSE there respective session_id32 again for subsequent key
+ *  generation sessions. If Alice and Bob fail to complete this session or
+ *  start a new session to generate a new key, they must NOT REUSE their
+ *  respective session_id32 again, but instead generate a new one. It is
+ *  recommended to always choose session_id32 uniformly at random to avoid
+ *  their reuse.
+ *
+ *  Returns: 0 if the arguments are invalid, 1 otherwise
+ *  Args:            ctx: pointer to a context object initialized for
+ *                        verification
+ *  Out:  vss_commitment: the coefficient commitments. The length of this array
+ *                        must be equal to the threshold (can be NULL).
+ *                 share: pointer to the key generation share
+ *   In:    session_id32: a 32-byte session_id32 as explained above
+ *               keypair: pointer to a keypair that contains the secret that is
+ *                        shared
+ *                    pk: pointer to the public key of the share recipient
+ *             threshold: the minimum number of signers required to produce a
+ *                        signature
+ */
+SECP256K1_API int secp256k1_frost_share_gen(
+    const secp256k1_context *ctx,
+    secp256k1_pubkey *vss_commitment,
+    secp256k1_frost_share *share,
+    const unsigned char *session_id32,
+    const secp256k1_keypair *keypair,
```

Comment:
Is it a good idea to pass the "secret that is shared" as input? As this secret shouldn't be used anywhere else because its no longer your secret after you shared it between n other people (assuming t < n)
Meaning that if a user uses his regular private key here he actually allows any t parties here to steal his money, which is unrelated to the generated public key at the end of the protocol.

Reply:
I think we need some sort of secret value passed to this API because, as far as I understand it, the secp APIs don't provide entropy directly (e.g. from urandom) but require that entropy to provided as an input.

It's a good call-out though, perhaps some strongly worded warnings and instructions can warn users not to use that secret for anything else. Do you have something else in mind? Is there a way we can avoid passing the secret that is shared as input?

Reply:
We can replace it with a 32 byte ephemeral(document that it has to be ephemeral) random entropy and also generate the first coefficient from chacha.
But maybe this limits some use-cases I'm not seeing

Reply:
The availability of the keypair for the first coefficient makes it [easy](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/138/files#diff-2d1c5749d9c83eb8473d43bed1b741f624529dd944ce21fc52614ac7e4bd4921R147) to reuse this keypair for the signed VSS hash that also is serving as a proof of knowledge.

However, as we discussed above, we might want to opt for a more conservative approach of sending the proof of knowledge in the first round independently of the signed VSS hash. If we do that, then `share_gen` can export a proof of knowledge, so the availability of the keypair for the first coefficient is not needed.

We'll still need some keypair to use for the signed VSS hash. Perhaps there should be a set of authentication public keys exchanged between the participants that can be used for signing the VSS hash and also for encrypting/decrypting shares.